libyul/optimiser/ValueConstraintBasedSimplifier.cpp

```diff
@@ -0,0 +1,394 @@
+		return ValueConstraint{}; // overflow
+	else
+		return bitsFromValue(
+			minValue + _other.minValue,
+			maxValue + _other.maxValue
+		);
+}
+
+ValueConstraint ValueConstraint::operator-(ValueConstraint const& _other)
+{
+	if (minValue < _other.maxValue)
+		return ValueConstraint{}; // underflow
+	else
+		return bitsFromValue(
+			minValue - _other.maxValue,
+			maxValue - _other.minValue
+		);
+}
+
+ValueConstraint ValueConstraint::operator<(ValueConstraint const& _other)
+{
+	if (maxValue < _other.minValue)
+		return constant(1);
+	else if (minValue >= _other.maxValue)
+		return constant(0);
+	else
+		return boolean();
+}
+
+ValueConstraint ValueConstraint::operator==(ValueConstraint const& _other)
+{
+	if (minValue == maxValue && minValue == _other.minValue && _other.minValue == _other.maxValue)
+		return constant(1);
+	else if (minBits == maxBits && minBits == _other.maxBits && _other.minBits == _other.maxBits)
+		return constant(1);
+	else if (maxValue < _other.minValue || minValue > _other.maxValue)
+		return constant(0);
+	else if ((maxBits == 0 && _other.minBits == 1) || (minBits == 1 && _other.maxBits == 0))
+		return constant(0);
+	else
+		return boolean();
+}
+
+ValueConstraint ValueConstraint::operator<<(ValueConstraint const& _other)
+{
+	if (boost::optional<u256> c = _other.isConstant())
+	{
+		if (*c >= 256)
+			return constant(0);
+		unsigned amount = unsigned(*c);
+		return ValueConstraint{
+			dev::bigintShiftLeftWorkaround(minValue, amount),
+			dev::bigintShiftLeftWorkaround(maxValue, amount),
+			dev::bigintShiftLeftWorkaround(minBits, amount),
+			dev::bigintShiftLeftWorkaround(maxBits, amount)
+		};
+	}
+	else
+	{
+		size_t lowestBit = lowestBitSet(maxBits);
+		if (lowestBit + bigint(_other.minValue) > 255)
+			return constant(0);
+		else
+			/// TODO could be more accurate, also taking "lastBit set" and _other.maxValue into account.
+			return bitRange(255, lowestBit + size_t(_other.minValue));
+	}
+}
+
+ValueConstraint ValueConstraint::operator>>(ValueConstraint const& _other)
+{
+	if (boost::optional<u256> c = _other.isConstant())
+	{
+		if (*c >= 256)
+			return constant(0);
+		unsigned amount = unsigned(*c);
+		return ValueConstraint{
+			minValue >> amount,
+			maxValue >> amount,
+			minBits >> amount,
+			maxBits >> amount
+		};
+	}
+	else
+	{
+		int highestBit = highestBitSet(maxBits);
+		if (highestBit - bigint(_other.minValue) < 0)
+			return constant(0);
+		else
+			/// TODO could be more accurate, also taking "lastBit set" and _other.maxValue into account.
+			return bitRange(size_t(highestBit - _other.minValue), 0);
+	}
+}
+
+boost::optional<u256> ValueConstraint::isConstant() const
+{
+	if (minValue == maxValue)
+		return minValue;
+	else
+		return {};
+}
+
+void ValueConstraintBasedSimplifier::visit(Expression& _expression)
+{
+	ASTModifier::visit(_expression);
+
+	// TODO this runs constraint deduction multiple times on the same sub-expression.
```

Comment:
How does it run multiple times?

Reply:
ASTModifier::visit calls back here for all sub-expressions.

Reply:
I don't have a solution to this. Do we want to go with it as it is or solve this before merging?